elfpy/markets/hyperdrive/hyperdrive_market.py

```diff
@@ -642,3 +640,169 @@ def apply_checkpoint(self, checkpoint_time: float, share_price: float) -> float:
             )
             self.market_state.apply_delta(market_deltas)
         return self.market_state.checkpoints[mint_time].share_price
+
+    def redeem_withdraw_shares(
+        self,
+        agent_wallet: wallet.Wallet,
+        shares: float,
+        min_output: float,
+        as_underlying: bool,
+    ) -> float:
+        r"""
+        Redeems withdrawal shares if enough margin has been freed to do so.
+
+        Parameters
+        ----------
+        agent_wallet: wallet.Wallet
+            The agent's wallet
+        shares: float
+            The withdrawal shares to redeem.
+        min_output: float
+            The minimum amount of base the LP expects to receive.
+        as_underlying: bool
+            If true, the user is paid in underlying, if false the contract transfers in yield source
+            directly. Note - for some paths one choice may be disabled or blocked.
+
+        Returns
+        -------
+        float
+            The amount of base the LP received.
+        """
+
+        # create a new checkpoint if necessary, close positions at the checkpoint time one
+        # position_duration ago.
+        self.apply_checkpoint(self.latest_checkpoint_time, self.market_state.share_price)
+        market_deltas, wallet_deltas = self.calc_redeem_withdraw_shares(shares, min_output, as_underlying)
+        self.update_market(market_deltas)
+        agent_wallet.update(wallet_deltas)
+
+        return wallet_deltas.balance.amount
+
+    def calc_redeem_withdraw_shares(
+        self, shares: float, min_output: float, as_underlying: bool
+    ) -> tuple[hyperdrive_actions.MarketDeltas, wallet.Wallet]:
+        r"""
+        Calculates the market and wallet deltas for redeemable withdrawal shares, if enough margin
+        has been freed to do so.
+
+        Parameters
+        ----------
+        shares: float
+            The withdrawal shares to redeem.
+        min_output: float
+            The minimum amount of base the LP expects to receive.
+        as_underlying: bool
+            If true, the user is paid in underlying, if false the contract transfers in yield source
+            directly. Note - for some paths one choice may be disabled or blocked.
+
+        Returns
+        -------
+        tuple[hyperdrive_actions.MarketDeltas, wallet.Wallet]
+
+        """
+        market_deltas = hyperdrive_actions.MarketDeltas()
```

Comment:
I'm thinking this should all be done with Decimal to match our use of it in the other `calc_` functions? Of course we might change it in a follow-up, but might as well stick to the pattern for now. This will help identify where changes need to be made later anyway.